backends/ebpf/psa/externs/ebpfPsaRegister.cpp

```diff
@@ -186,17 +189,24 @@ void EBPFRegisterPSA::emitRegisterRead(CodeBuilder *builder, const P4::ExternMet
     builder->emitIndent();
 
     if (leftExpression != nullptr) {
-        codeGen->visit(leftExpression);
-        builder->append(" = ");
-        if (this->initialValue != nullptr) {
-            builder->append(this->initialValue->value.str());
-        } else {
-            builder->append("(");
+        if (initialValue != nullptr || leftExpression->type->is<IR::Type_Bits>()) {
+            // let's create fake assigment statement and use it to generate valid code
+            const IR::Expression *right =
+                initialValue != nullptr ? initialValue : new IR::Constant(leftExpression->type, 0);
+            const auto *assigment = new IR::AssignmentStatement(leftExpression, right);
+            ControlBodyTranslatorPSA cg(*translator);
+            assigment->apply(cg);
```

Comment:
does this work?
For example, the typemap won't always have a type for the right.
In general, it would be nice to do this in a separate pass before code generation.

Reply:
I think it should works (PTF tests are passing, especially [`register.RegisterDefaultPSATest`](https://github.com/p4lang/p4c/blob/21105c8ea809ee255464784d0a671dd58f567868/backends/ebpf/tests/ptf/register.py#L84-L98)).
If the initial value is defined for register, then type map contains type for right - because of expression generated by frontend.
If the initial value is not defined, type map does not contain type for right, but type is copied from left. For this case exist this line: https://github.com/p4lang/p4c/blob/7554cb932dfae7db0b72a08c3747f1008b27bfa6/backends/ebpf/codeGen.cpp#L449

Maybe it is not the best way, but is there any reason to not to use `expression->type`?

Reply:
Expression->type is only populated if you run the type inference with a flag indicating that you want this.
Some simple expressions like constants may have the type correct, though.
If it works, let's leave it like this.